src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -136,58 +150,123 @@ class NodeSyncConnection {
-  /// When the periodical timer wake up, it'll check whether these two fields are set
-  /// and it'll only send data when these are set.
-  std::vector<RaySyncMessages *> responses_;
-  std::vector<grpc::ServerUnaryReactor *> unary_reactors_;
+  void StartPull() {
+    receiving_message_ = std::make_shared<RaySyncMessage>();
+    RAY_LOG(DEBUG) << "Start reading: " << NodeID::FromBinary(GetRemoteNodeID());
+    StartRead(receiving_message_.get());
+  }
+
+  /// grpc requests for sending and receiving
+  std::shared_ptr<const RaySyncMessage> sending_message_;
+  std::shared_ptr<RaySyncMessage> receiving_message_;
 };
 
-/// SyncConnection for gRPC client side. It has customized logic for sending.
-class ClientSyncConnection : public NodeSyncConnection {
+/// Reactor for gRPC server side. It has customized logic for sending.
+class RayServerBidiReactor : public BidiReactor<ServerBidiReactor> {
  public:
-  ClientSyncConnection(
+  RayServerBidiReactor(
+      grpc::CallbackServerContext *server_context,
       instrumented_io_context &io_context,
-      const std::string &node_id,
-      std::function<void(std::shared_ptr<RaySyncMessage>)> message_processor,
-      std::shared_ptr<grpc::Channel> channel);
+      const std::string &local_node_id,
+      std::function<void(std::shared_ptr<const RaySyncMessage>)> message_processor,
+      std::function<void(const std::string &, bool)> cleanup_cb);
+
+  ~RayServerBidiReactor() override = default;
+
+  void Disconnect() override;
 
  protected:
-  /// Send the message from the pending queue to the target node.
-  /// It'll use gRPC to send the message directly.
-  void DoSend() override;
+  void OnCancel() override;
+  void OnDone() override;
 
-  /// Start to send long-polling request to remote nodes.
-  void StartLongPolling();
+  /// grpc callback context
+  grpc::CallbackServerContext *server_context_;
+};
 
-  /// Stub for this connection.
-  std::unique_ptr<ray::rpc::syncer::RaySyncer::Stub> stub_;
+/// Reactor for gRPC client side. It has customized logic for sending.
+class RayClientBidiReactor : public BidiReactor<ClientBidiReactor> {
+ public:
+  RayClientBidiReactor(
+      const std::string &remote_node_id,
+      const std::string &local_node_id,
+      instrumented_io_context &io_context,
+      std::function<void(std::shared_ptr<const RaySyncMessage>)> message_processor,
+      std::function<void(const std::string &, bool)> cleanup_cb,
+      std::unique_ptr<ray::rpc::syncer::RaySyncer::Stub> stub);
+
+  ~RayClientBidiReactor() override = default;
 
-  /// Dummy request for long-polling.
-  DummyRequest dummy_;
+  void Disconnect() override;
+
+ protected:
+  /// Callback from gRPC
+  void OnDone(const grpc::Status &status) override;
+
+  /// grpc callback context
+  grpc::ClientContext client_context_;
+
+  /// grpc requests for sending and receiving
+  std::shared_ptr<const RaySyncMessage> sending_message_;
```

Comment:
hmm why does this class has private members (sending_message_, receiving_message_) which also defined in its' parent?

Reply:
This is good catch.